File: dotnet/CarePlatform.Models/Common/TiuDocument.cs
using CarePlatform.Models.Common;

namespace CarePlatform.Models.Common;

/// <summary>
/// Shared TIU document record parsed from TIU DOCUMENTS BY CONTEXT RPC.
/// Used by Notes (class=3), DCSumm (class=244), and Consults TIU sub-documents.
///
/// RPC returns 15 caret-delimited pieces (from cprs/uDocTree.pas):
///   1  - Document IEN
///   2  - Document Title
///   3  - FM date of document
///   4  - Patient Name
///   5  - DUZ;Author name  (format: "DUZ;Initials;Display Name")
///   6  - Location
///   7  - Status
///   8  - ADM/VIS: date;FMDate
///   9  - Discharge Date;FMDate
///   10 - Package variable pointer
///   11 - Number of images
///   12 - Subject
///   13 - Has children  (+, >, <, %, combinations)
///   14 - Parent document (IEN or context number for top-level)
///   15 - Order children of ID Note by title rather than date
/// </summary>
public class TiuDocument
{
    public string Ien { get; set; } = "";
    public string Title { get; set; } = "";
    public string FmDate { get; set; } = "";
    public DateTime? ReferenceDate { get; set; }
    public string PatientName { get; set; } = "";

    /// <summary>Raw DUZ;Initials;DisplayName from piece 5.</summary>
    public string AuthorRaw { get; set; } = "";
    /// <summary>Author display name extracted from piece 5 (third semicolon-delimited part).</summary>
    public string AuthorName { get; set; } = "";
    /// <summary>Author DUZ extracted from piece 5 (first semicolon-delimited part).</summary>
    public long AuthorDuz { get; set; }

    public string LocationName { get; set; } = "";
    public string Status { get; set; } = "";

    /// <summary>ADM/VIS: date;FMDate string from piece 8.</summary>
    public string VisitString { get; set; } = "";
    /// <summary>Discharge Date;FMDate string from piece 9.</summary>
    public string DischargeDateString { get; set; } = "";
    public DateTime? DischargeDate { get; set; }

    /// <summary>Package variable pointer (e.g., "GMR(123,456" for consults) from piece 10.</summary>
    public string PackageRef { get; set; } = "";
    public int ImageCount { get; set; }
    public string Subject { get; set; } = "";

    /// <summary>
    /// Has children flag from piece 13.
    /// Values: +  = has children, > = is ID child, &lt; = is ID parent,
    ///         % = group/category node, combinations like +>, +&lt;, *prefix.
    /// Empty = leaf document with no children.
    /// </summary>
    public string HasChildrenFlag { get; set; } = "";
    public bool HasChildren => !string.IsNullOrEmpty(HasChildrenFlag);

    /// <summary>Parent document IEN from piece 14. For top-level notes, equals the context number (e.g., "1").</summary>
    public string ParentDocument { get; set; } = "";

    /// <summary>If true, order children of this ID Note by title rather than date (piece 15).</summary>
    public bool OrderByTitle { get; set; }

    /// <summary>
    /// Parse a single caret-delimited line from TIU DOCUMENTS BY CONTEXT RPC.
    /// Field indices match cprs/uDocTree.pas MakeNoteTreeObject (pieces 1-15, 0-based array 0-14).
    /// </summary>
    public static TiuDocument Parse(string vistaLine)
    {
        var p = VistaStringParser.Split(vistaLine);
        var authorRaw = p.Length > 4 ? p[4] : "";   // piece 5: "DUZ;Initials;DisplayName"
        var authorParts = authorRaw.Split(';');
        var hasChildrenRaw = p.Length > 12 ? p[12] : "";
        // Strip leading '*' from hasChildren per Pascal: if Copy(DocHasChildren, 1, 1) = '*' then ...
        if (hasChildrenRaw.StartsWith('*'))
            hasChildrenRaw = hasChildrenRaw[1..];

        var dischargeDateStr = p.Length > 8 ? p[8] : "";  // piece 9: "Dis: date;FMDate"
        // Extract FM date portion from discharge date string (after semicolon)
        var dischargeFmDate = dischargeDateStr.Contains(';')
            ? dischargeDateStr.Split(';')[1]
            : dischargeDateStr;

        return new TiuDocument
        {
            Ien = p.Length > 0 ? p[0] : "",
            Title = p.Length > 1 ? p[1] : "",
            FmDate = p.Length > 2 ? p[2] : "",
            ReferenceDate = p.Length > 2 ? VistaStringParser.ParseFmDateTime(p[2]) : null,
            PatientName = p.Length > 3 ? p[3] : "",
            AuthorRaw = authorRaw,
            AuthorDuz = authorParts.Length > 0 ? long.TryParse(authorParts[0], out var duz) ? duz : 0 : 0,
            AuthorName = authorParts.Length > 2 ? authorParts[2] : (authorParts.Length > 1 ? authorParts[1] : authorRaw),
            LocationName = p.Length > 5 ? p[5] : "",
            Status = p.Length > 6 ? p[6] : "",
            VisitString = p.Length > 7 ? p[7] : "",
            DischargeDateString = dischargeDateStr,
            DischargeDate = VistaStringParser.ParseFmDateTime(dischargeFmDate),
            PackageRef = p.Length > 9 ? p[9] : "",
            ImageCount = p.Length > 10 ? (int.TryParse(p[10], out var ic) ? ic : 0) : 0,
            Subject = p.Length > 11 ? p[11] : "",
            HasChildrenFlag = hasChildrenRaw,
            ParentDocument = p.Length > 13 ? p[13] : "",
            OrderByTitle = p.Length > 14 && p[14] == "1"
        };
    }

    /// <summary>
    /// Parse a list of caret-delimited lines from TIU DOCUMENTS BY CONTEXT RPC.
    /// Skips blank lines and lines with no IEN.
    /// </summary>
    public static List<TiuDocument> ParseList(List<string> vistaLines)
    {
        return vistaLines
            .Where(l => !string.IsNullOrWhiteSpace(l))
            .Select(Parse)
            .Where(d => !string.IsNullOrEmpty(d.Ien))
            .ToList();
    }
}

File: python/app/models/tiu_document.py
# Copyright (c) 2026 Engineered Care, Inc. Licensed under the MIT License.
"""TIU document model — mirrors TiuDocument.cs.

Parsed from TIU DOCUMENTS BY CONTEXT RPC (15 caret-delimited pieces).
Used by Notes (class=3), DCSumm (class=244), and Consults TIU sub-documents.
"""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel

from app.models.vista_string_parser import parse_fm_datetime, parse_int, split


class TiuDocument(BaseModel):
    ien: str = ""
    title: str = ""
    fm_date: str = ""
    reference_date: Optional[datetime] = None
    patient_name: str = ""
    author_raw: str = ""
    author_name: str = ""
    author_duz: int = 0
    location_name: str = ""
    status: str = ""
    visit_string: str = ""
    discharge_date_string: str = ""
    discharge_date: Optional[datetime] = None
    package_ref: str = ""
    image_count: int = 0
    subject: str = ""
    has_children_flag: str = ""
    has_children: bool = False
    parent_document: str = ""
    order_by_title: bool = False

    @classmethod
    def parse(cls, vista_line: str) -> "TiuDocument":
        """Parse a single caret-delimited line from TIU DOCUMENTS BY CONTEXT."""
        p = split(vista_line)
        author_raw = p[4] if len(p) > 4 else ""
        author_parts = author_raw.split(";")

        has_children_raw = p[12] if len(p) > 12 else ""
        if has_children_raw.startswith("*"):
            has_children_raw = has_children_raw[1:]

        discharge_date_str = p[8] if len(p) > 8 else ""
        discharge_fm = (
            discharge_date_str.split(";")[1]
            if ";" in discharge_date_str
            else discharge_date_str
        )

        author_duz = 0
        if author_parts and author_parts[0]:
            try:
                author_duz = int(author_parts[0])
            except ValueError:
                pass

        if len(author_parts) > 2:
            author_name = author_parts[2]
        elif len(author_parts) > 1:
            author_name = author_parts[1]
        else:
            author_name = author_raw

        return cls(
            ien=p[0] if len(p) > 0 else "",
            title=p[1] if len(p) > 1 else "",
            fm_date=p[2] if len(p) > 2 else "",
            reference_date=parse_fm_datetime(p[2]) if len(p) > 2 else None,
            patient_name=p[3] if len(p) > 3 else "",
            author_raw=author_raw,
            author_duz=author_duz,
            author_name=author_name,
            location_name=p[5] if len(p) > 5 else "",
            status=p[6] if len(p) > 6 else "",
            visit_string=p[7] if len(p) > 7 else "",
            discharge_date_string=discharge_date_str,
            discharge_date=parse_fm_datetime(discharge_fm),
            package_ref=p[9] if len(p) > 9 else "",
            image_count=parse_int(p[10]) if len(p) > 10 else 0,
            subject=p[11] if len(p) > 11 else "",
            has_children_flag=has_children_raw,
            has_children=bool(has_children_raw),
            parent_document=p[13] if len(p) > 13 else "",
            order_by_title=(p[14] == "1") if len(p) > 14 else False,
        )

    @classmethod
    def parse_list(cls, vista_lines: list[str]) -> list["TiuDocument"]:
        """Parse a list of lines, skipping blanks and IEN-less entries."""
        results = []
        for line in vista_lines:
            if not line or not line.strip():
                continue
            doc = cls.parse(line)
            if doc.ien:
                results.append(doc)
        return results
